fossas/broker
# Ane 1060 broker smart imports

# Overview

- Broker is able to toggle if users want to scan branches, as well as being able to target specific branches to scan
- Broker is able to toggle if users want to scan tags

### Update Import Criteria
- If configured, import the latest HEAD from configured branches
  - Primary branch is `main`, or `master` if that doesn't exist
  - Users may modify the list of branches, in which case they must re-specify the primary branch
      - As in, no configured branches means 'try to figure out priary and import it', while configured branches mean 'this is the complete list of branches to import'
      - The absence of a primary branch should be an error, but absence of configured branches should be warnings; they should be able to configured for a 'best effort' import
      - Support glob matches for branches, e.g. 'import branches that match `release-candidate/*` 
   - If configured, import the latest tag

```
integrations:
- type: git
  import_branches: true  <- new, defaults to true
  watched_branches:      <- new, if unspecified defaults to "figure out primary";
    - main                  if specified but import_branches is false, warn that
    - release/*             this config is contradictory
  import_tags: true      <- new, defaults to false
  poll_interval: 1h
  url: git@github.com:fossas/broker.git
  auth:
    type: ssh_key_file
    path: /home/me/.ssh/id_rsa
```

### Temporal Config Trackign
- If `import_branches` goes from `true`  to `false`, branch state tracked in the DB should be deleted, so that if it is ever flipped to `true` again this counts as a 'first time' import of tags
- If `import_tags` goes from `true`  to `false`, branch state tracked in the DB should be deleted, so that if it is ever flipped to `true` again this counts as a 'first time' import of branches

## Acceptance criteria

- The new import criteria is in place
- A warning is raised to users when running Broker for the first time if it was previously run on an older version
- The `import_branches` and `import_tags` items are temporally compared, and changes handled on startup

## Testing plan

- `test_integration_smart_imports_default` (config.rs)
    - ensure that if import_tags and import_branches is not set then import_tags is set to `false` , import_branches is set to 'true' 
- `test_integration_smart_imports_malformed`
   - ensure error is populated when import_branches is `false` and watched_branches is configured
- `remove_all_branch_states` (sqlite.rs)
    - ensure that the query to remove all branches for an integration is successful
- `remove_all_tag_states` (sqlite.rs)
    - ensure that the query to remove all tags for an integration is successful

- Manually tested that initial run with new broker version
    - there were no breaking changes
    - users are warned that they are on an older version of broker based on version in the db and broker applies the new db changes
- Manually tested import_branches set to `true` and watched_branches is empty
    - Watched branches will be set to main/master if that is present 
    - If not main/master in the repo, then broker will throw an error
- Manually tested toggle import_branches from `true` to `false`
    - Looked at the database for the scanned entries for the branches for the integration
    - After setting the field to `false` looked into the db and saw that all the branches for the integration were deleted
    - ran cargo run with import_branches set to `true` and the branches were scanned as if it were it's first time
- Manually tested toggle import_tags from `true` to `false`
    - Looked at the database for the scanned entries for the tags for the integration
    - After setting the field to `false` looked into the db and saw that all the tags for the integration were deleted
    - ran cargo run with import_tags set to `true` and the tags were scanned as if it were it's first time
- Manually tested glob support feature
    - configured watched_branches to include [test*] on a repo that had branches [main, test-branch-1, test-branch-2]
        - broker scanned branches [test-branch-1, test-branch-2]
 

## Risks

_Highlight any areas that you're unsure of, want feedback on, or want reviewers to pay particular attention to._

_Example: I'm not sure I did X correctly, can reviewers please double-check that for me?_

## References

_Add links to any referenced GitHub issues, Zendesk tickets, Jira tickets, Slack threads, etc._

_Example:_

- [ANE-1060](https://fossa.atlassian.net/browse/ANE-1060)
- Zendesk Ticket IDs
   - 6524, 6652, 6685, 6765

## Checklist



[ANE-1060]: https://fossa.atlassian.net/browse/ANE-1060?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
